### ISSUE TITLE
GPIO Binding not working

### ISSUE THREAD

**Issue by berho** [openhab/openhabian#39]
Thirst of all thanks for the great work on openHABian! It is so easy to set up an raspberry pi with openHAB now :)

The main reason I use a raspberry is the GPIO board to connect to the real world ;) but the GPIO binding does not work on openHAB. I tried to find a solution and was spending a few days with this. The workarounds found did not work for me:
* run openHAB as root: no-go for security reason and dont't know how to do
* add user to group "gpio": no such group on openHABian
* install special versions of libraries: didn't try, but the reason why I use openHABian is to quickly install and use openHAB. If i have to manually install things after the setup I don't really need openHABian at all

**Comment by ThomDietrich** [comment]
I never used the gpio binding.
I see people asking about the gpio binding a lot and this is not a problem with openHABian but rather with the tight constraints+requirements the binding defines. Did you consider using wiringPi in conjunction with the exec binding? I have a similar usecase and am very happy with wiringPi. Random Tutorial: https://projects.drogon.net/raspberry-pi/gpio-examples/tux-crossing/gpio-examples-1-a-single-led

**Comment by bernhard-a** [comment]
My project is running fine with openHABian and the GPIO binding now but only when I start openHAB2 with "sudo /etc/init.d/openhab2 stop" and "sudo /usr/share/openhab2/start.sh".

Output and input GPIOs are working as expected. But I have to do this manually and I don't want to run openHAB2 with root privileges in production environment. 

Does anybody know how to add the group "gpio" to the system and the GPIO ports?

**Comment by ThomDietrich** [comment]
Okay great, your system is a base for further testing to figure out the **right** solution.
Please first have a look at [this and following comments](https://community.openhab.org/t/migration-to-openhab2/17469/7?u=thomdietrich).

Instead of manually starting as root you can set root as the go-to user in `/usr/lib/systemd/system/openhab2.service`. Please try this at least once to evaluate the approach.

However as you said, it would be preferable to not execute openhab as root. We need to add the ``gpio`` group and add user ``openhab`` to it. The group itself has to have control over the gpios. Please implement what can be found as the desired state in [this thread](https://www.raspberrypi.org/forums/viewtopic.php?f=31&t=77742). The important steps seem to be

1. create the ``gpio`` group
2. add ``pi`` and `openhab` to `gpio`
3. `chgrp -R gpio /sys/class/gpio`
4. maybe adapt group permissions on these files

Another option would be to use pigpio. It's started as a daemon and thereafter you can access the gpios (via the daemon) without special privileges. It is another download though and probably not compatible with the gpio binding. Maybe it should be?

**Comment by bernhard-a** [comment]
I have testet 
- set root as go-to user: work 
- add pi and openhab to gpio group: I get "[ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item MyItemName, exception: /sys/class/gpio/export"

Any ideas?

**Comment by ThomDietrich** [comment]
what does the following say? 
```
ls - la /sys/class/gpio/export
ls - la /sys/class/gpio
```

**Comment by bernhard-a** [comment]
```
[19:59:41] pi@openHABianPi:~$ ls -la /sys/class/gpio/export
--w------- 1 root root 4096 Dec  4 21:20 /sys/class/gpio/export
[19:59:54] pi@openHABianPi:~$ ls -la /sys/class/gpio
total 0
drwxr-xr-x  2 root root    0 Dec  5 20:00 .
drwxr-xr-x 48 root root    0 Jan  1  1970 ..
--w-------  1 root root 4096 Dec  4 21:20 export
lrwxrwxrwx  1 root root    0 Dec  4 21:20 gpio17 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpio17
lrwxrwxrwx  1 root root    0 Dec  4 21:20 gpio18 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpio18
lrwxrwxrwx  1 root root    0 Dec  4 21:20 gpio21 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpio21
lrwxrwxrwx  1 root root    0 Dec  4 21:20 gpio22 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpio22
lrwxrwxrwx  1 root root    0 Dec  4 21:20 gpio4 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpio4
lrwxrwxrwx  1 root root    0 Jan  1  1970 gpiochip0 -> ../../devices/platform/soc/3f200000.gpio/gpio/gpiochip0
lrwxrwxrwx  1 root root    0 Jan  1  1970 gpiochip100 -> ../../devices/platform/soc/soc:virtgpio/gpio/gpiochip100
--w-------  1 root root 4096 Jan  1  1970 unexport
```

It seems that the group is set to root every reboot :(

**Comment by ThomDietrich** [comment]
Seems like I'll have to dissect a normal raspbian setup!
So do I understand you correctly, it works after setting the owner group to gpio and before rebooting?

**Comment by bernhard-a** [comment]
No, does not work after setting owner group before reboot too.

**Comment by skatun** [comment]
can you explain this in more detail?
Instead of manually starting as root you can set root as the go-to user in /usr/lib/systemd/system/openhab2.service. Please try this at least once to evaluate the approach.

on my rpi its an empty file..

**Comment by ThomDietrich** [comment]
Depends on your system. With openHABian the systemd service management system is used. If your system is a bit older and is based on init.d, you need to have a look at "/etc/init.d/openhab2", which uses ``USER_AND_GROUP``. You should probably add this to `/etc/default/openhab2`:

```
USER_AND_GROUP=root:root
```

**Comment by skatun** [comment]
well i have the openhabian version 2.0 build 624, but the service file is empty.. so i should jus add USER_AND_GROUP=root:root to it?

**Comment by skatun** [comment]
my output for your reference...
pi@openHABianPi:~$ ls -la /sys/class/gpio
total 0
drwxr-xr-x  2 root gpio    0 Dec  7 21:25 .
drwxr-xr-x 46 root root    0 Jan  1  1970 ..
-rwxrwxr-x  1 root gpio 4096 Dec  4 19:32 export
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio1 -> ../../devices/platform/soc/20                                                                                        200000.gpio/gpio/gpio1
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio10 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio10
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio11 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio11
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio12 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio12
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio13 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio13
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio15 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio15
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio16 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio16
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio17 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio17
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio18 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio18
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio19 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio19
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio20 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio20
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio21 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio21
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio22 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio22
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio23 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio23
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio24 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio24
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio25 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio25
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio26 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio26
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio27 -> ../../devices/platform/soc/2                                                                                        0200000.gpio/gpio/gpio27
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio3 -> ../../devices/platform/soc/20                                                                                        200000.gpio/gpio/gpio3
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio4 -> ../../devices/platform/soc/20                                                                                        200000.gpio/gpio/gpio4
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio5 -> ../../devices/platform/soc/20                                                                                        200000.gpio/gpio/gpio5
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio6 -> ../../devices/platform/soc/20                                                                                        200000.gpio/gpio/gpio6
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio7 -> ../../devices/platform/soc/20                                                                                        200000.gpio/gpio/gpio7
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio8 -> ../../devices/platform/soc/20                                                                                        200000.gpio/gpio/gpio8
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpio9 -> ../../devices/platform/soc/20                                                                                        200000.gpio/gpio/gpio9
lrwxrwxrwx  1 root gpio    0 Dec  7 21:24 gpiochip0 -> ../../devices/platform/so                                                                                        c/20200000.gpio/gpio/gpiochip0
-rwxrwxr-x  1 root gpio 4096 Nov 29 20:51 unexport

**Comment by ThomDietrich** [comment]
Do not mix "openHAB" and "openHABian".
If you are using openHABian, you should see a file ``/usr/lib/systemd/system/openhab2.service`` where you can set user and password. If the file does not exist, please check for the correct path by executing `sudo systemctl status openhab2`

Regarding your gpio: the listing looks good to me (except that you forgot to put it in code fences ;) )

**Comment by skatun** [comment]
I dont get any errors either from the log:
`23:11:15.910 [INFO ] [marthome.event.ItemStateChangedEvent] - Light_Module2 changed from ON to OFF
`
my item:
`Switch Light_Module2 		"Module 2" 		(Group_LivingRoom,Lights,Group_LightLogging) { gpio="pin:19" }  //Relay 15 `

but relay does not switch on:(

Does running as root work?

**Comment by skatun** [comment]
some output from the log:
```
23:34:52.894 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Contact_BedroomL, exception: /sys/class/gpio/export
23:34:52.912 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Heating_Bathroom, exception: /sys/class/gpio/export
23:34:52.949 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Light_Module1, exception: /sys/class/gpio/export
23:34:52.979 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Light_Module2, exception: /sys/class/gpio/export
23:34:53.000 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Light_Module3, exception: /sys/class/gpio/export
23:34:53.019 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Light_Module4, exception: /sys/class/gpio/export
23:34:53.098 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Contact_BathroomR, exception: /sys/class/gpio/export
23:34:53.125 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Div_DoorOpener, exception: /sys/class/gpio/export
23:34:53.161 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Heating_Kitchen, exception: /sys/class/gpio/export
23:34:53.190 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Light_Bathroom, exception: /sys/class/gpio/export
23:34:53.219 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Light_Kitchen, exception: /sys/class/gpio/export
23:34:53.248 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Light_MasterBedroom, exception: /sys/class/gpio/export
23:34:53.279 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Contact_BedroomR, exception: /sys/class/gpio/export
23:34:53.300 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Contact_BalconyL, exception: /sys/class/gpio/export
23:34:53.340 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Contact_BalconyR, exception: /sys/class/gpio/export
23:34:53.380 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Heating_LivingRoom, exception: /sys/class/gpio/export
23:34:53.412 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Heating_MasterBedroom, exception: /sys/class/gpio/export
23:34:53.451 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Light_Balcony, exception: /sys/class/gpio/export
23:34:53.490 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Contact_BathroomL, exception: /sys/class/gpio/export
23:34:53.518 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Heating_Hallway, exception: /sys/class/gpio/export
23:34:53.571 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Heating_Bedroom, exception: /sys/class/gpio/export
23:34:53.599 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Light_Hallway, exception: /sys/class/gpio/export
23:34:53.639 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Contact_LivingRoomR, exception: /sys/class/gpio/export
23:34:53.663 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Contact_MasterBedroomL, exception: /sys/class/gpio/export
23:34:53.709 [ERROR] [ab.binding.gpio.internal.GPIOBinding] - Error occured while creating backend object for item Contact_LivingRoomL, exception: /sys/class/gpio/export
23:34:53.838 [ERROR] [org.apache.felix.configadmin        ] - [org.osgi.service.cm.ManagedService, id=258, bundle=176/mvn:org.openhab.io/org.openhab.io.gpio/1.9.0-SNAPSHOT]: Unexpected problem updating configuration org.openhab.gpio
java.lang.ClassCastException: java.lang.Boolean cannot be cast to java.lang.String
        at org.openhab.io.gpio.linux.GPIOLinux.updated(GPIOLinux.java:126)[176:org.openhab.io.gpio:1.9.0.201611290210]
        at org.apache.felix.cm.impl.helper.ManagedServiceTracker.updated(ManagedServiceTracker.java:189)[7:org.apache.felix.configadmin:1.8.8]
        at org.apache.felix.cm.impl.helper.ManagedServiceTracker.updateService(ManagedServiceTracker.java:152)[7:org.apache.felix.configadmin:1.8.8]
        at org.apache.felix.cm.impl.helper.ManagedServiceTracker.provideConfiguration(ManagedServiceTracker.java:85)[7:org.apache.felix.configadmin:1.8.8]
        at org.apache.felix.cm.impl.ConfigurationManager$ManagedServiceUpdate.provide(ConfigurationManager.java:1444)[7:org.apache.felix.configadmin:1.8.8]
        at org.apache.felix.cm.impl.ConfigurationManager$ManagedServiceUpdate.run(ConfigurationManager.java:1400)[7:org.apache.felix.configadmin:1.8.8]
        at org.apache.felix.cm.impl.UpdateThread.run0(UpdateThread.java:143)[7:org.apache.felix.configadmin:1.8.8]
        at org.apache.felix.cm.impl.UpdateThread.run(UpdateThread.java:110)[7:org.apache.felix.configadmin:1.8.8]
        at java.lang.Thread.run(Thread.java:745)[:1.8.0_111]
```

**Comment by bernhard-a** [comment]
Running as root work as expected, but I think it is not a good idea. It is a security risk!

**Comment by EliasGabrielsson** [comment]
I think these lines could fix GPIO access:
```
# Config GPIO acces
sudo adduser openhab gpio
sudo usermod -a -G dialout openhab
```
I can't verifiy this until end of December when I have my pi next to me.
@bernhard-a could you test this on a fresh install?
@ThomDietrich This could be a solution for this issue.

**Comment by ThomDietrich** [comment]
@EliasGabrielsson this will not work... :/ The problem is, that there is no gpio group nor will it have the right privileges if you create it by yourself. Adding openhab to grou gpio would be the [right next step](https://github.com/openhab/openhabian/issues/39#issuecomment-264148885).

Your second line is not needed, that's already part of your openHABian setup ;)

**Comment by dochego** [comment]
I have just installed the March release of openhabian on a RP3, and found that
- group gpio exists "out-of-the-box" and gives the right access privileges to the gpio pins
- user openhabian is a member of the gpio group, but openhab is not, so openhab fails to access the GPIO.

All I had to do was "sudo adduser openhab gpio" (as suggested by EliasGabrielsson) to get the GPIO working with openhab.

Could this be implemented in the default setup (i.e. add user openhab to group gpio)?
